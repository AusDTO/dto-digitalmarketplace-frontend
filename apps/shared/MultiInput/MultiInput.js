/* eslint-disable */

import React from 'react'
import PropTypes from 'prop-types'

class MultiInput extends React.Component {
  static propTypes = {
    rows: PropTypes.array,
    defaultRows: PropTypes.number,
    onChange: PropTypes.func,
    name: PropTypes.string.isRequired
  }

  static defaultProps = {
    onChange: () => {},
    onBlur: () => {},
    onFocus: () => {}
  }

  constructor(props) {
    super(props)

    let inputs
    if (props.rows && props.rows.length) {
      inputs = props.rows.map((row, i) => this.createRow(i, row))
    }

    this.state = {
      inputs: inputs || Array.from({ length: props.defaultRows || 1 }).map((_, i) => this.createRow(i))
    }
  }

  /**
   * Pad out new row
   * @param  {number}
   * @param  {String}
   * @return {object}
   */
  createRow(id: number, value: string = '') {
    return { id, value }
  }

  /**
   * Get an array of values from the inputs.
   * Removes empty values
   * @return {array}
   */
  getValues() {
    return this.state.inputs.map(input => input.value).filter(i => i)
  }

  onChange(id: number, e: ReactEvent) {
    const { onChange } = this.props

    // @see https://fb.me/react-event-pooling
    e.persist()

    this.setState(
      previousState => {
        let newInputs = previousState.inputs.map(input => {
          if (input.id === id) {
            input.value = e.target.value
          }

          return input
        })

        return { inputs: newInputs }
      },
      () => onChange(this.getValues())
    )
  }

  addRow(e: ReactEvent) {
    e.preventDefault()
    const { inputs } = this.state
    const lastRow = inputs[inputs.length - 1]

    this.setState({
      inputs: inputs.concat(this.createRow(lastRow.id + 1))
    })
  }

  removeRow(id: number, e: ReactEvent) {
    e.preventDefault()
    const { inputs } = this.state
    const { onChange } = this.props

    this.setState(
      {
        inputs: inputs.filter(r => r.id !== id)
      },
      () => onChange(this.getValues())
    )
  }

  render() {
    const { inputs } = this.state
    const { name, onBlur, onFocus, className = '', describedby, hint } = this.props
    return (
      <div>
        {inputs.map(({ id, value }, i) => {
          let fieldName = `${name}[]`
          let fieldId = `${name}-${i}`
          return (
            <div key={id} className="list-entry row">
              <div className="col-xs-12 col-sm-9">
                <label htmlFor={fieldId} className="text-box-number-label">
                  {i + 1}.
                </label>
                <input
                  type="text"
                  name={fieldName}
                  id={fieldId}
                  onChange={this.onChange.bind(this, id)}
                  onBlur={onBlur}
                  onFocus={onFocus}
                  className={`text-box ${className}`}
                  aria-describedby={hint && describedby}
                  defaultValue={value}
                  className="au-text-input au-text-input--block multiInput"
                />
              </div>
              {i > 0 &&
                <a href="#" onClick={this.removeRow.bind(this, id)}>
                  remove <span className="visuallyhidden">number {i + 1}</span>
                </a>}
            </div>
          )
        })}
        <a href="#" onClick={this.addRow.bind(this)}>
          Add another row
        </a>
      </div>
    )
  }
}

export default MultiInput
