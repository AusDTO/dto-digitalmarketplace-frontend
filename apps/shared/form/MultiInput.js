/* eslint-disable */

import React from 'react'
import PropTypes from 'prop-types'
import { Control, controls } from 'react-redux-form'

import StatefulError from './StatefulError'
import MultiInputComponent from 'shared/MultiInput'

const MultiInput = props => {
  const { id, name, htmlFor, label, description, model, messages, controlProps, mapProps, validators } = props

  return (
    <div className="field">
      <label htmlFor={htmlFor} className="question-heading au-text-input__label">
        {label}
      </label>
      {description &&
        <div className="au-text-input__hint" id={`${id}-hint`}>
          {description}
        </div>}
      {messages && <StatefulError model={model} messages={messages} id={id} />}
      <Control
        model={model}
        controlProps={{ name, id, describedby: `${id}-hint`, hint: description, ...controlProps }}
        mapProps={{
          rows: ({ viewValue }) => viewValue,
          className: ({ fieldValue }) => (!fieldValue.valid && fieldValue.touched ? 'invalid' : ''),
          ...mapProps,
          ...controls.default
        }}
        validators={validators}
        component={MultiInputComponent}
      />
    </div>
  )
}

MultiInput.defaultProps = {
  mapProps: {}
}

MultiInput.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  htmlFor: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  model: PropTypes.oneOfType([PropTypes.func, PropTypes.string]).isRequired,

  messages: PropTypes.object,
  validators: PropTypes.object,
  description: PropTypes.string,
  controlProps: PropTypes.object,
  mapProps: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
}

export default MultiInput
