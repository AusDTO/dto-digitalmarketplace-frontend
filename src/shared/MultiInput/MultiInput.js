import React from 'react'

class MultiInput extends React.Component {
  static propTypes = {
    rows: React.PropTypes.array,
    defaultRows: React.PropTypes.number,
    onChange: React.PropTypes.func,
    name: React.PropTypes.string.isRequired
  }

  static defaultProps = {
    onChange: () => {},
    onBlur: () => {},
    onFocus: () => {},
  }

  constructor(props) {
    super(props);

    let inputs;
    if (props.rows) {
      inputs = props.rows.map((row, i) => this.createRow(i, row))
    }

    this.state = {
      inputs: inputs || Array
        .from({ length: props.defaultRows || 1 })
        .map((_, i) => this.createRow(i))
    };
  }

  /**
   * Pad out new row
   * @param  {number}
   * @param  {String}
   * @return {object}
   */
  createRow(id: number, value: string = '') {
    return { id, value }
  }

  /**
   * Get an array of values from the inputs.
   * Removes empty values
   * @return {array}
   */
  getValues() {
    return this.state.inputs.map(input => input.value).filter(i => i);
  }

  onChange(id: number, e: ReactEvent) {
    const { onChange } = this.props;

    // @see https://fb.me/react-event-pooling
    e.persist();

    this.setState((previousState) => {
      let newInputs = previousState.inputs.map(input => {
        if (input.id === id) {
          input.value = e.target.value;
        }

        return input;
      });

      return { inputs: newInputs };
    }, () => onChange(this.getValues()));

  }

  addRow(e: ReactEvent) {
    e.preventDefault();
    const { inputs } = this.state;
    const lastRow = inputs[inputs.length - 1];

    this.setState({
      inputs: inputs.concat(this.createRow(lastRow.id + 1))
    });

  }

  removeRow(id: number, e: ReactEvent) {
    e.preventDefault()
    const { inputs } = this.state;
    const { onChange } = this.props;

    this.setState({
      inputs: inputs.filter(r => r.id !== id)
    }, () => onChange(this.getValues()));
  }

  render() {
    const { inputs } = this.state;
    const { name, onBlur, onFocus, className = '' } = this.props;
    return (
      <div>
        {inputs.map(({ id, value }, i) => {
          let fieldName = `${name}[]`;
          let fieldId = i > 0 ? `${name}-${i}` : name
          return (
            <div key={id} className="list-entry">
              <label htmlFor={fieldId} className="text-box-number-label">{i + 1}</label>
              <input
                type="text"
                name={fieldName}
                id={fieldId}
                onChange={this.onChange.bind(this, id)}
                onBlur={onBlur}
                onFocus={onFocus}
                className={`text-box ${className}`}
                defaultValue={value} />
              {i > 0 && (
                <button type="button" className="button-secondary" onClick={this.removeRow.bind(this, id)}>
                  remove <span className="visuallyhidden">number {i + 1}</span>
                </button>
              )}
            </div>
          )
        })}
        <button type="button" className="button-secondary" onClick={this.addRow.bind(this)}>
          Add another row
        </button>
      </div>
    )
  }
}

export default MultiInput;