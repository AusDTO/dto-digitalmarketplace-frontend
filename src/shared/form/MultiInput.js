import React from 'react';
import { Control, controls } from 'react-redux-form';

import StatefulError from './StatefulError';
import MultiInputComponent from '../MultiInput';

const MultiInput = (props) => {
  const {
    id,
    name,
    htmlFor,
    label,
    description,
    model,
    messages,
    controlProps,
    mapProps,
    validators
  } = props;

  return (
    <div className="field">
      <label htmlFor={htmlFor}>{label}</label>
      {description && (
        <p className="hint" id={`${id}-hint`}>{description}</p>
      )}
      {messages &&
        <StatefulError
          model={model}
          messages={messages}
          id={id}
        />
      }
      <Control
        model={model}
        controlProps={{ name, id, describedby: `${id}-hint`, hint: description, ...controlProps}}
        mapProps={{
          rows: ({ viewValue }) => viewValue,
          className: ({ fieldValue }) => !fieldValue.valid && fieldValue.touched ? 'invalid' : '',
          ...mapProps,
          ...controls.default,
        }}
        validators={validators}
        component={MultiInputComponent}
      />
    </div>
  );
};

MultiInput.defaultProps = {
  mapProps: {},
};

MultiInput.propTypes = {
  id: React.PropTypes.string.isRequired,
  name: React.PropTypes.string.isRequired,
  htmlFor: React.PropTypes.string.isRequired,
  label: React.PropTypes.string.isRequired,
  model: React.PropTypes.oneOfType([
    React.PropTypes.func,
    React.PropTypes.string,
  ]).isRequired,

  messages: React.PropTypes.object,
  validators: React.PropTypes.object,
  description: React.PropTypes.string,
  controlProps: React.PropTypes.object,
  mapProps: React.PropTypes.oneOfType([
    React.PropTypes.func,
    React.PropTypes.object,
  ]),
}

export default MultiInput;