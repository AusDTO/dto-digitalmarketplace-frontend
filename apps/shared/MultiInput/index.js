import MultiInput from './MultiInput'

export { MultiInput as default, MultiInput }
