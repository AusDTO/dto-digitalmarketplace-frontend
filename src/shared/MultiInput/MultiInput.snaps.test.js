import React from 'react'
import renderer from 'react-test-renderer'

import MultiInput from './MultiInput'

test('MultiInput with no props', () => {
  const component = renderer.create(
    <MultiInput name="test" />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('MultiInput with 4 rows', () => {
  const rows = [ 'One', 'Two', 'Three', 'Four' ];
  const component = renderer.create(
    <MultiInput name="test" rows={rows} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('MultiInput with 4 default rows', () => {
  const component = renderer.create(
    <MultiInput name="test" defaultRows={4} />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
