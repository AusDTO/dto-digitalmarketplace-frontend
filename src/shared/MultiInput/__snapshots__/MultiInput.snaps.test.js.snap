exports[`test MultiInput with 4 default rows 1`] = `
<div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test">
      1
      .
    </label>
    <input
      className="text-box "
      defaultValue=""
      id="test"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
  </div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test-1">
      2
      .
    </label>
    <input
      className="text-box "
      defaultValue=""
      id="test-1"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
    <button
      className="button-secondary"
      onClick={[Function]}
      type="button">
      remove 
      <span
        className="visuallyhidden">
        number 
        2
      </span>
    </button>
  </div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test-2">
      3
      .
    </label>
    <input
      className="text-box "
      defaultValue=""
      id="test-2"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
    <button
      className="button-secondary"
      onClick={[Function]}
      type="button">
      remove 
      <span
        className="visuallyhidden">
        number 
        3
      </span>
    </button>
  </div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test-3">
      4
      .
    </label>
    <input
      className="text-box "
      defaultValue=""
      id="test-3"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
    <button
      className="button-secondary"
      onClick={[Function]}
      type="button">
      remove 
      <span
        className="visuallyhidden">
        number 
        4
      </span>
    </button>
  </div>
  <button
    className="button-secondary"
    onClick={[Function]}
    type="button">
    Add another row
  </button>
</div>
`;

exports[`test MultiInput with 4 rows 1`] = `
<div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test">
      1
      .
    </label>
    <input
      className="text-box "
      defaultValue="One"
      id="test"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
  </div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test-1">
      2
      .
    </label>
    <input
      className="text-box "
      defaultValue="Two"
      id="test-1"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
    <button
      className="button-secondary"
      onClick={[Function]}
      type="button">
      remove 
      <span
        className="visuallyhidden">
        number 
        2
      </span>
    </button>
  </div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test-2">
      3
      .
    </label>
    <input
      className="text-box "
      defaultValue="Three"
      id="test-2"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
    <button
      className="button-secondary"
      onClick={[Function]}
      type="button">
      remove 
      <span
        className="visuallyhidden">
        number 
        3
      </span>
    </button>
  </div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test-3">
      4
      .
    </label>
    <input
      className="text-box "
      defaultValue="Four"
      id="test-3"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
    <button
      className="button-secondary"
      onClick={[Function]}
      type="button">
      remove 
      <span
        className="visuallyhidden">
        number 
        4
      </span>
    </button>
  </div>
  <button
    className="button-secondary"
    onClick={[Function]}
    type="button">
    Add another row
  </button>
</div>
`;

exports[`test MultiInput with no props 1`] = `
<div>
  <div
    className="list-entry">
    <label
      className="text-box-number-label"
      htmlFor="test">
      1
      .
    </label>
    <input
      className="text-box "
      defaultValue=""
      id="test"
      name="test[]"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text" />
  </div>
  <button
    className="button-secondary"
    onClick={[Function]}
    type="button">
    Add another row
  </button>
</div>
`;
